### PR TITLE
eth/tracers: improve tracing performance

### PR DESCRIPTION
**Pull request by holiman**
This PR first adds a benchmark of the "opcode tracer", i.e, the barebones tracer which stores data in memory before sending it out over RPC. The benchmark is a very simple loop which just runs through `1M` gas. 

The second commit adds some optimizations which improves the performance quite a lot: 
```
name                old time/op    new time/op    delta
TransactionTrace-6     100ms ±12%      28ms ±13%  -72.03%  (p=0.000 n=10+9)

name                old alloc/op   new alloc/op   delta
TransactionTrace-6    40.1MB ± 4%     9.0MB ±12%  -77.58%  (p=0.000 n=10+10)

name                old allocs/op  new allocs/op  delta
TransactionTrace-6      571k ± 0%      163k ± 0%  -71.42%  (p=0.000 n=10+8)
```

### CODE REVIEW

**Comment by holiman** [comment]
Further improvements: 
```
name                old time/op    new time/op    delta
TransactionTrace-6    28.1ms ±13%    22.7ms ± 8%  -19.07%  (p=0.000 n=9+9)

name                old alloc/op   new alloc/op   delta
TransactionTrace-6    8.98MB ±12%    7.14MB ±14%  -20.55%  (p=0.000 n=10+10)

name                old allocs/op  new allocs/op  delta
TransactionTrace-6      163k ± 0%       82k ± 0%  -49.98%  (p=0.002 n=8+10)
```
Total: 
```
name                old time/op    new time/op    delta
TransactionTrace-6     100ms ±12%      23ms ± 8%  -77.36%  (p=0.000 n=10+9)

name                old alloc/op   new alloc/op   delta
TransactionTrace-6    40.1MB ± 4%     7.1MB ±14%  -82.18%  (p=0.000 n=10+10)

name                old allocs/op  new allocs/op  delta
TransactionTrace-6      571k ± 0%       82k ± 0%  -85.71%  (p=0.000 n=10+10)
```

**Comment by emadami3060** [comment]
25800

در تاریخ سه‌شنبه ۸ ژوئن ۲۰۲۱،‏ ۱۶:۲۴ Martin Holst Swende <
***@***.***> نوشت:

**Comment by holiman** [comment]
I ran this against [this](https://etherscan.io/tx/0x87ba373aa69ee675ae764b49c3d179d2367dd0d970c36b9177dc67f00603b4e9) transaction, 

```
#!/bin/bash

time ./build/bin/geth attach --exec "debug.traceTransaction('0x87ba373aa69ee675ae764b49c3d179d2367dd0d970c36b9177dc67f00603b4e9', {'disableMemory':true, 'disableStorage':true})" 
```
It took about `52s`, and `27s` with stack disabled. 
I also took a cpu profile when running it with stack enabled: 
profile001

**Review by chfast — APPROVED** [review]

**Comment by MariusVanDerWijden** [comment]
So I tried this on a really heavy tx we have on ropsten
` time ./build/bin/geth --ropsten attach --exec "debug.traceTransaction('0xcafaf3669df640365da47527902e3e4dff48bd334ab0c5c894dae5ba1f82df27', {'disableMemory':true, 'disableStorage':true, reexec:10000})"`

This PR: 
(60 seconds of it is state regeneration)
```
Executed in  281.69 secs    fish           external
   usr time  151.47 secs  323.00 micros  151.47 secs
   sys time  103.44 secs   67.00 micros  103.44 secs
```

Current Master: 
(59 seconds of it is state regeneration)
(I killed the node, call didn't finish)
```
Executed in  269.17 mins    fish           external
   usr time    6.21 secs  358.00 micros    6.21 secs
   sys time    0.50 secs   77.00 micros    0.50 secs
```

**Comment by holiman** [comment]
> So I tried this on a really heavy tx we have on ropsten

Looks like it worked then? Can the PR haz an approve?

File: core/vm/logger.go [hunk 1]
```diff
@@ -238,7 +246,7 @@ func WriteTrace(writer io.Writer, logs []StructLog) {
 		if len(log.Stack) > 0 {
 			fmt.Fprintln(writer, "Stack:")
 			for i := len(log.Stack) - 1; i >= 0; i-- {
-				fmt.Fprintf(writer, "%08d  %x\n", len(log.Stack)-i-1, math.PaddedBigBytes(log.Stack[i], 32))
+				fmt.Fprintf(writer, "%08d  %v\n", len(log.Stack)-i-1, log.Stack[i].Hex())
```
Comment (fjl) [inline core/vm/logger.go:249]: You can use %s here.

Comment (holiman) [reply core/vm/logger.go]: done!

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

**Comment by holiman** [comment]
> 0xcafaf3669df640365da47527902e3e4dff48bd334ab0c5c894dae5ba1f82df27

18M gas on that transactions -- yeah might be a bit slow :)

**Pull request merged**